### ISSUE TITLE
FastAPI integration Async Redis session dependency override, error during test RuntimeError: <Queue > is bound to a different event loop

### ISSUE THREAD

**Issue by christiansicari** [cunla/fakeredis-py#292]
**Describe the bug**
I am currently developing a FastAPI app that gets a Redis Session from a Dependency Injection, and uses it to write some data on Redis. 
I need to test the endpoint, therefore I created a pytext fixture which in turn:
1. create a FakeAsyncRedis
2. override the dependency in the endpoint
3. return the same FakeAsyncRedis object

in the test basically I call the endpoint, and I read the key I expect to get set, but I get the following error
`test_redis - RuntimeError: <Queue at 0x111538a10 maxsize=0 tasks=1> is bound to a different event loop`

**To Reproduce**
My fixtures are:

```python
@pytest.fixture(scope="session")
def event_loop():
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()
    yield loop
    loop.close()


@pytest.fixture
async def redis(event_loop: asyncio.AbstractEventLoop):
    redis_session = fakeredis.FakeAsyncRedis(decode_responses=True, encoding="utf-8", )

    async def _get_async_redis_session() -> AsyncIterator[fakeredis.FakeAsyncRedis]:
        yield redis_session

    for app in all_fast_apps:
        app.dependency_overrides[get_async_redis_session] = _get_async_redis_session
    return redis_session
```
This is my dependency used to fetch a Redis Connection:
```python
from redis import asyncio as aioredis
async def get_async_redis_session() -> AsyncIterator[Redis]:
    async with aioredis.from_url(redisUrl, **redis_args) as session:
        yield session
    await session.close()
```

and this is the endpoint definition:
```python
async def write_on_redis(
    request: Request,
    redis: Redis = Depends(get_async_redis_session)
) :
  await redis.set(name=key, value=value, ex=expire)
```
Finally the test where I get the error:

```python
def test_write(redis):  # redis is the fixture we defined before

       url = "myapiendpoint"
        response = client.get(url)
       value = await redis.get(key) <-- this throws the error
```
**Expected behavior**
Get the key set before in the endpoint 

**Desktop (please complete the following information):**
 - fakeredis==2.21.1
 - redis==4.6.0

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/cunla) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/cunla/fakeredis-py/issues/292">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/cunla/fakeredis-py/issues/292/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/cunla/fakeredis-py/issues/292/pledge.svg">
</picture>
</a>

**Comment by csicari-voxy** [comment]
@cunla thanks for reading :)
Sure, this is the stack trace
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.pyenv/versions/3.11.5/envs/li/lib/python3.11/site-packages/redis/asyncio/client.py:513: in execute_command
    conn = self.connection or await pool.get_connection(command_name, **options)
../../.pyenv/versions/3.11.5/envs/li/lib/python3.11/site-packages/redis/asyncio/connection.py:1381: in get_connection
    if await connection.can_read_destructive():
../../.pyenv/versions/3.11.5/envs/li/lib/python3.11/site-packages/redis/asyncio/connection.py:759: in can_read_destructive
    return await self._parser.can_read_destructive()
../../.pyenv/versions/3.11.5/envs/li/lib/python3.11/site-packages/redis/asyncio/connection.py:250: in can_read_destructive
    return await self._stream.read(1)
../../.pyenv/versions/3.11.5/envs/li/lib/python3.11/site-packages/fakeredis/aioredis.py:83: in read
    return await self._socket.responses.get()  # type:ignore
../../.pyenv/versions/3.11.5/lib/python3.11/asyncio/queues.py:155: in get
    getter = self._get_loop().create_future()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Queue at 0x1152f8510 maxsize=0 tasks=1>

    def _get_loop(self):
        loop = events._get_running_loop()
    
        if self._loop is None:
            with _global_lock:
                if self._loop is None:
                    self._loop = loop
        if loop is not self._loop:
>           raise RuntimeError(f'{self!r} is bound to a different event loop')
E           RuntimeError: <Queue at 0x1152f8510 maxsize=0 tasks=1> is bound to a different event loop
```
Actually, I have same output with and without the event_loop fixture

**Comment by cunla** [comment]
I don't think this is related to FakeRedis..

See this thread:
https://stackoverflow.com/questions/53724665/using-queues-results-in-asyncio-exception-got-future-future-pending-attached

I suspect you are creating your FastAPI server and the fakeredis connection on different event-loops

**Comment by cunla** [comment]
Here you are creating a new event-loop:

```python
@pytest.fixture(scope="session")
def event_loop():
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()
    yield loop
    loop.close()
```

**Comment by csicari-voxy** [comment]
It's the same as I do for Postgres actually. I create a new event loop with that fixture and I share it between Fastapi and Postgres, just passing the fixture. I think will dig into that later, now I need a fast workaround. I'll keep here update if you prefer :)

**Comment by cunla** [comment]
Please update here. 
If you are running into this, I am sure others do as well :)

**Comment by steve-mavens** [comment]
Indeed I have also hit this problem. I couldn't figure out how to get it all working. You'd think that `event_loop` fixture should be unnecessary, but it's a workaround for a different issue where sometimes if you don't create your own, `pytest_asyncio` ends up closing the event loop at the "wrong" time, and you get warnings from destructors that things are not as they should be. I can't remember what happened when I tried making the `redis` fixture depend on the custom `event_loop` fixture to straighten out the order everything was created.

Anyway in the end I worked around it using `mock.patch` (of a function called by my fastapi dependency function) instead of the dependency override.

**Comment by steve-mavens** [comment]
I should probably warn that this code is a bit scrappy. It doesn't bother undoing the modification to `app.dependency_overrides`, which happens to work with my code because every test that uses the app at all, uses it via this fixture and wants the same dependency override. It wouldn't work if I had different tests doing different things with that other dependency.

**Comment by steve-mavens** [comment]
I can add something about how to intersect fastapi and async fakeredis :-) I'm making some assumptions about how the fastapi app is set up, and I don't know how typical my code is. I'm not deeply involved in the fastapi community to know what's most common: I just got the dependencies I needed working and then mostly forgot about them.

**Comment by cunla** [comment]
Do you mind reviewing [this section](https://fakeredis.readthedocs.io/en/latest/#use-to-test-fastapi) in the documentation and see whether it makes sense. If so, close this issue please

**Comment by steve-mavens** [comment]
I don't think it's complete. The code you've used doesn't explain what `create_backend` is, so I don't think there's enough information there for a fastapi user to adapt the code snippet to their app. The reason I haven't done anything on this yet is that it's a fair bit of work to produce a minimal example, figure out under what circumstances the "different event loop" problem actually bites, and therefore what the best fix is to advise. What worked for me is the combination of avoiding `app.dependency_overrides` and writing all the tests as async.

**Comment by cunla** [comment]
Well, I'll keep this open, if any of you can spend some time to improve the docs it would be great. I haven't worked with FastAPI so I am not in a position to write it.

**Comment by steve-mavens** [comment]
I've just updated one of my projects to the latest pytest-asyncio (released March 19th), and at least for my usage the need to have a session-scoped event loop has gone away. I can just remove it and I don't get the "bound to a different event loop" errors any more.

The reason I even made the experiment of removing it is that pytest-asyncio has deprecated user-redefining the `event_loop` fixture, and I wanted to make that warning go away :-). I haven't investigated what the actual change is in pytest-asyncio that improved the cleanup such that it's no longer needed.

It might be that this is only fixed because all my other fixtures are function-scoped as well, but as a brief experiment I did also get the tests working by setting some of my fixtures to `scope="session"`, and *also* using `pytest.mark.asyncio(scope="session")`. So I'm cautiously optimistic that it could also work now if there were some reason the fakeredis fixture did need to be session-scoped.

@christiansicari would it be possible for you to check this for your usage? That is:
* update to pytest-asyncio v.0.23.6
* if you used my hack then change the fixture back to setting the fastapi dependency override
* set scope="session" on the fixture
* set `pytest.mark.asyncio(scope="session")` on all the the tests that use this fixture.

Then see whether the event loop errors have gone.

Hopefully tomorrow or the day after I will get to my project from which I took the code snippet above, and can test the fastapi dependency override for myself.

**Comment by sjjessop** [comment]
@cunla I have created a standalone FastAPI app that uses Redis, and written a test for it to demonstrate the dependency override. Then I've copied that code into your docs and the PR is here: https://github.com/cunla/fakeredis-py/pull/311

This does not need the eventloop fixture like in @christiansicari 's original question. I believe this to be due to improvements in pytest-asyncio since this question was asked, since my project used to need the same thing and no longer does.

**Comment by sjjessop** [comment]
Something is very wrong with that PR. For me the page now just shows a massive flickering discord URL. I have no idea how, and I'm now going to check from another machine. Could be some kind of compromise. Anyway, here is an alternative PR that should be equivalent: https://github.com/cunla/fakeredis-py/pull/312

**Comment by sjjessop** [comment]
This issue is that a user named "simulified" posted a comment to the PR with content that (when Javascript is enabled) replaces the whole content of the page with an image. I've reported the user to github, but I don't know if that will have any effect. I have also blocked the user, although that doesn't prevent their comments appearing for me. When I load the page with javascript disabled, I can see the malicious comment, but github has decided in its wisdom that the "..." menu on comments should only work with Javascript enabled. So, I can't even try to delete the comment.

The other PR is OK for now, but this feels like the kind of exploit that could hit evey repo on github within hours until they deal with it...
